cocotb/binary.py

```diff
@@ -32,28 +32,61 @@
 
 import os
 import random
+import re
 import warnings
 
+
+_RESOLVE_TO_0      = "-lL"
+_RESOLVE_TO_1      = "hH"
+_RESOLVE_TO_CHOICE = "xXzZuUwW"
 resolve_x_to = os.getenv('COCOTB_RESOLVE_X', "VALUE_ERROR")
 
-def resolve(string):
-    for char in BinaryValue._resolve_to_0:
-        string = string.replace(char, "0")
-    for char in BinaryValue._resolve_to_1:
-        string = string.replace(char, "1")
-    for char in BinaryValue._resolve_to_error:
-        if resolve_x_to == "VALUE_ERROR" and char in string:
-            raise ValueError("Unable to resolve to binary >%s<" % string)
+
+class ResolveTable(dict):
+    """Translation table class for resolving binary strings.
+
+    For use with :func:`str.translate()`, which indexes into table with Unicode ordinals.
+    """
+    def __init__(self):
+        self.update({ord("0"): ord("0"), ord("1"): ord("1")})
+        self.update({ord(k): ord("0") for k in _RESOLVE_TO_0})
+        self.update({ord(k): ord("1") for k in _RESOLVE_TO_1})
+
+        # Do not resolve if resolve_x_to is not set to one of the supported values
+        def no_resolve(key):
+            return key
+        self.resolve_x = no_resolve
+
+        if resolve_x_to == "VALUE_ERROR":
+            def resolve_error(key):
+                raise ValueError("Unresolvable bit in binary string: '{}'".format(chr(key)))
+            self.resolve_x = resolve_error
         elif resolve_x_to == "ZEROS":
-            string = string.replace(char, "0")
+            self.update({ord(k): ord("0") for k in _RESOLVE_TO_CHOICE})
         elif resolve_x_to == "ONES":
-            string = string.replace(char, "1")
+            self.update({ord(k): ord("1") for k in _RESOLVE_TO_CHOICE})
         elif resolve_x_to == "RANDOM":
-            bits = "{0:b}".format(random.getrandbits(1))
-            string = string.replace(char, bits)
-    return string
+            def resolve_random(key):
+                # convert to correct Unicode ordinal:
+                # ord('0') = 48
+                # ord('1') = 49
+                return random.getrandbits(1) + 48
+            self.resolve_x = resolve_random
+
+        self._resolve_to_choice = {ord(c) for c in _RESOLVE_TO_CHOICE}
+
+    def __missing__(self, key):
+        if key in self._resolve_to_choice:
+            return self.resolve_x(key)
```

Comment:
This still won't actually do what you're expecting it to, `str.translate` caches the calls to `__missing__`.

Reply:
All instances of each character to be translated will be the same, either `0` or `1`, which is the same as current behavior. I noted this in the commit message.